File: backend/src/test/java/org/globe42/web/persons/PersonControllerMvcTest.java
package org.globe42.web.persons;

import static org.globe42.test.Answers.modifiedFirstArgument;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.time.LocalDate;
import java.util.Collections;
import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.PersonDao;
import org.globe42.domain.Person;
import org.globe42.test.GlobeMvcTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC test for {@link PersonController}
 * @author JB Nizet
 */
@RunWith(SpringRunner.class)
@GlobeMvcTest(PersonController.class)
public class PersonControllerMvcTest {

    @MockBean
    private PersonDao mockPersonDao;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private MockMvc mvc;

    private Person person;

    @Before
    public void prepare() {
        person = new Person(1L);
        person.setAdherent(true);
        person.setEntryDate(LocalDate.of(2017, 5, 21));
        person.setMediationCode("A2");
    }

    @Test
    public void shouldList() throws Exception {
        when(mockPersonDao.findNotDeleted()).thenReturn(Collections.singletonList(person));

        mvc.perform(get("/api/persons"))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(1));
    }

    @Test
    public void shouldListDeleted() throws Exception {
        when(mockPersonDao.findDeleted()).thenReturn(Collections.singletonList(person));

        mvc.perform(get("/api/persons").param("deleted", ""))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(1));
    }

    @Test
    public void shouldGet() throws Exception {
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));

        mvc.perform(get("/api/persons/{personId}", person.getId()))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$.id").value(1))
           .andExpect(jsonPath("$.entryDate").value("2017-05-21"))
           .andExpect(jsonPath("$.entryDate").value("2017-05-21"));
    }

    @Test
    public void should404IfPersonNotFound() throws Exception {
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.empty());

        mvc.perform(get("/api/persons/{personId}", person.getId()))
           .andExpect(status().isNotFound());
    }

    @Test
    public void shouldCreate() throws Exception {
        when(mockPersonDao.save(any(Person.class))).thenAnswer(modifiedFirstArgument((Person p) -> p.setId(1L)));

        mvc.perform(post("/api/persons")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsBytes(PersonControllerTest.createCommand())))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.id").value(1));
    }

    @Test
    public void shouldUpdate() throws Exception {
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));

        mvc.perform(put("/api/persons/{personId}", person.getId())
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsBytes(PersonControllerTest.createCommand())))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldDelete() throws Exception {
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));

        mvc.perform(delete("/api/persons/{personId}", person.getId()))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldResurrect() throws Exception {
        when(mockPersonDao.findById(person.getId())).thenReturn(Optional.of(person));

        mvc.perform(delete("/api/persons/{personId}/deletion", person.getId()))
           .andExpect(status().isNoContent());
    }
}

File: backend/src/test/kotlin/org/globe42/web/persons/PersonControllerMvcTest.kt
package org.globe42.web.persons

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.globe42.dao.CountryDao
import org.globe42.dao.CoupleDao
import org.globe42.dao.MembershipDao
import org.globe42.dao.PersonDao
import org.globe42.domain.Country
import org.globe42.domain.Gender
import org.globe42.domain.PARIS_TIME_ZONE
import org.globe42.domain.Participation
import org.globe42.domain.Person
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.*
import java.time.LocalDate

/**
 * MVC test for [PersonController]
 * @author JB Nizet
 */
@GlobeMvcTest(PersonController::class)
@MockkBean(CoupleDao::class)
class PersonControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {

    @MockkBean
    private lateinit var mockPersonDao: PersonDao

    @MockkBean
    private lateinit var mockCountryDao: CountryDao

    @MockkBean
    private lateinit var mockMembershipDao: MembershipDao

    private lateinit var person: Person

    @BeforeEach
    fun prepare() {
        person = Person(1L, "John", "Doe", Gender.MALE).apply {
            entryDate = LocalDate.of(2017, 5, 21)
            mediationCode = "A2"
        }

        every { mockCountryDao.findByIdOrNull(any()) } returns Country("FRA", "France")
    }

    @Test
    fun `should list`() {
        every { mockPersonDao.findNotDeleted() } returns listOf(person)

        mvc.get("/api/persons").andExpect {
            status { isOk() }
            jsonValue("$[0].id", 1)
        }
    }

    @Test
    fun `should list deleted`() {
        every { mockPersonDao.findDeleted() } returns listOf(person)

        mvc.get("/api/persons") {
            param("deleted", "")
        }.andExpect {
            status { isOk() }
            jsonValue("$[0].id", 1)
        }
    }

    @Test
    fun `should get`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person

        mvc.get("/api/persons/{personId}", person.id).andExpect {
            status { isOk() }
            jsonValue("$.id", 1)
            jsonValue("$.entryDate", "2017-05-21")
            jsonValue("$.entryDate", "2017-05-21")
        }
    }

    @Test
    fun `should 404 if person not found`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns null

        mvc.get("/api/persons/{personId}", person.id).andExpect {
            status { isNotFound() }
        }
    }

    @Test
    fun `should create`() {
        every { mockPersonDao.nextMediationCode(any()) } returns 1
        every { mockPersonDao.save(any<Person>()) } answers { arg<Person>(0).apply { id = 1L } }

        mvc.post("/api/persons") {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(PersonControllerTest.createCommand())
        }.andExpect {
            status { isCreated() }
            jsonValue("$.id", 1)
        }
    }

    @Test
    fun `should update`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person
        every { mockPersonDao.nextMediationCode(any()) } returns 1
        mvc.put("/api/persons/{personId}", person.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(PersonControllerTest.createCommand())
        }.andExpect {
            status { isNoContent() }
        }
    }

    @Test
    fun `should delete`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person

        mvc.delete("/api/persons/{personId}", person.id).andExpect {
            status { isNoContent() }
        }
    }

    @Test
    fun `should resurrect`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person

        mvc.delete("/api/persons/{personId}/deletion", person.id).andExpect {
            status { isNoContent() }
        }
    }

    @Test
    fun `should signal death`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person
        person.addParticipation(Participation(65L))

        val command = PersonDeathCommandDTO(LocalDate.now())
        mvc.put("/api/persons/{personId}/death", person.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isNoContent() }
        }

        assertThat(person.deathDate).isEqualTo(command.deathDate)
        assertThat(person.getParticipations()).isEmpty()
    }

    @Test
    fun `should get reminders`() {
        every { mockPersonDao.findByIdOrNull(person.id!!) } returns person
        every { mockMembershipDao.findByPersonAndYear(person, LocalDate.now(PARIS_TIME_ZONE).year) } returns null

        mvc.get("/api/persons/{personId}/reminders", person.id)
            .andExpect {
                status { isOk() }
                jsonValue("$[0].type", ReminderType.MEMBERSHIP_TO_RENEW.name)
            }
    }
}
